Inconsistency in handling the return of preprocess_ci!

https://github.com/JuliaDebug/Cthulhu.jl/blob/9fdc45c233cf593bc426fb6aea6f16095efe001d/src/Cthulhu.jl#L261-L262

returns `ci` but the later code uses `codeinf`. Is this deliberate? Note it's different from

https://github.com/JuliaDebug/Cthulhu.jl/blob/9fdc45c233cf593bc426fb6aea6f16095efe001d/test/runtests.jl#L28-L42

where the return value is used henceforth. The latter is similar to #177, which fixed some problems, and that leads me to think that the one in `src/Cthulhu.jl` must be wrong. But in fact if you use the one in `test/runtests.jl`, here's what you get:

```julia
ulia> @find_callsites_by_ftt tryparse(VersionNumber, "v2.1.3")
ERROR: AssertionError: argextype only works on argument-position values
Stacktrace:
 [1] argextype(x::Any, src::Core.Compiler.IRCode, sptypes::Vector{Any}, slottypes::Vector{Any})
   @ Core.Compiler ./compiler/utilities.jl:229
 [2] (::Cthulhu.var"#13#16"{Core.Compiler.IRCode, Vector{Any}, Vector{Any}})(arg::Expr)
   @ Cthulhu ~/.julia/dev/Cthulhu/src/reflection.jl:46
 [3] map!(f::Cthulhu.var"#13#16"{Core.Compiler.IRCode, Vector{Any}, Vector{Any}}, dest::Vector{Any}, A::Vector{Any})
   @ Base ./abstractarray.jl:2794
 [4] mapany(f::Function, A::Vector{Any})
   @ Base ./abstractarray.jl:2803
 [5] find_callsites(interp::Cthulhu.CthulhuInterpreter, CI::Core.Compiler.IRCode, stmt_info::Vector{Any}, mi::Core.MethodInstance, slottypes::Vector{Any}, optimize::Bool; params::Core.Compiler.NativeInterpreter)
   @ Cthulhu ~/.julia/dev/Cthulhu/src/reflection.jl:46
 [6] find_callsites
   @ ~/.julia/dev/Cthulhu/src/reflection.jl:28 [inlined]
 [7] find_callsites_by_ftt(f::Any, TT::Any; optimize::Bool)
   @ Main ~/.julia/dev/Cthulhu/test/runtests.jl:40
 [8] find_callsites_by_ftt(f::Any, TT::Any)
   @ Main ~/.julia/dev/Cthulhu/test/runtests.jl:38
 [9] top-level scope
   @ REPL[4]:1
```
Not sure if this is a Julia bug or something else.

--- a/src/Cthulhu.jl
+++ b/src/Cthulhu.jl
@@ -267,7 +267,7 @@ function _descend(term::AbstractTerminal, interp::CthulhuInterpreter, mi::Method
             else
                 (codeinf, rt, infos, slottypes) = lookup(interp, mi, optimize)
             end
-            ci = preprocess_ci!(codeinf, mi, optimize, CONFIG)
+            codeinf = preprocess_ci!(codeinf, mi, optimize, CONFIG)
             callsites = find_callsites(interp, codeinf, infos, mi, slottypes, optimize; params)
 
             if display_CI

--- a/src/reflection.jl
+++ b/src/reflection.jl
@@ -30,7 +30,7 @@ function find_callsites(interp::CthulhuInterpreter, CI::Union{Core.CodeInfo, IRC
     for (id, c) in enumerate(isa(CI, IRCode) ? CI.stmts.inst : CI.code)
         callsite = nothing
         isa(c, Expr) || continue
-        if stmt_info !== nothing
+        if stmt_info !== nothing && is_call_expr(c, optimize)
             info = stmt_info[id]
             if info !== nothing
                 rt = argextype(SSAValue(id), CI, sptypes, slottypes)
@@ -38,9 +38,7 @@ function find_callsites(interp::CthulhuInterpreter, CI::Union{Core.CodeInfo, IRC
                 # so extract rhs for such an case
                 args = c.args
                 if !optimize
-                    if isexpr(c, :(=))
-                        args = c.args[2].args
-                    end
+                    args = (ignorelhs(c)::Expr).args
                 end
                 types = mapany(arg -> widenconst(argextype(arg, CI, sptypes, slottypes)), args)
                 was_return_type = false
@@ -118,10 +116,8 @@ function find_callsites(interp::CthulhuInterpreter, CI::Union{Core.CodeInfo, IRC
         end
 
         if callsite === nothing && c isa Expr
-            if c.head === :(=)
-                c = c.args[2]
-                (c isa Expr) || continue
-            end
+            c = ignorelhs(c)
+            (c isa Expr) || continue
             if c.head === :invoke
                 rt = argextype(SSAValue(id), CI, sptypes, slottypes)
                 at = argextype(c.args[2], CI, sptypes, slottypes)
@@ -153,6 +149,12 @@ function find_callsites(interp::CthulhuInterpreter, CI::Union{Core.CodeInfo, IRC
         end
     end
     return callsites
+end
+
+ignorelhs(@nospecialize(x)) = isexpr(x, :(=)) ? last(x.args) : x
+function is_call_expr(x::Expr, optimize::Bool)
+    optimize && isexpr(x, :invoke) && return true
+    return isexpr(ignorelhs(x), :call)
 end
 
 function dce!(ci, mi)

--- a/test/runtests.jl
+++ b/test/runtests.jl
@@ -8,23 +8,6 @@ using Revise
 
 @test isempty(detect_ambiguities(Cthulhu))
 
-function firstassigned(specializations::Core.SimpleVector)
-    # the methodinstance may not be first (annoying)
-    for i = 1:length(specializations)
-        if isassigned(specializations, i)
-            return specializations[i]
-        end
-    end
-    return nothing
-end
-function firstassigned(specializations)
-    mis = []
-    Base.visit(specializations) do mi
-        isempty(mis) && push!(mis, mi)
-    end
-    return mis[1]
-end
-
 function process(@nospecialize(f), @nospecialize(TT=()); optimize=true)
     (interp, mi) = Cthulhu.mkinterp(f, TT)
     (ci, rt, infos, slottypes) = Cthulhu.lookup(interp, mi, optimize; allow_no_codeinf=true)
@@ -100,6 +83,14 @@ end
 
     # Note the failure of `callinfo` to properly handle specialization
     @test_broken Cthulhu.callinfo(Tuple{typeof(twice), AbstractFloat}, AbstractFloat) isa Cthulhu.MultiCallInfo
+end
+
+# https://github.com/JuliaDebug/Cthulhu.jl/issues/196
+let
+    callsites = @find_callsites_by_ftt tryparse(VersionNumber, "v2.1.3")
+    @test !isempty(callsites)
+    callsites = @find_callsites_by_ftt optimize=false tryparse(VersionNumber, "v2.1.3")
+    @test !isempty(callsites)
 end
 
 # Check that we see callsites that are the rhs of assignments

